Tribler/community/tunnel/tunnel_community.py

```diff
@@ -135,30 +135,48 @@ def enabled(self):
     def sendto(self, data, destination):
         if self.check_num_packets(destination, False):
             if TunnelConversion.is_allowed(data):
-                if dispersy.util.is_valid_address(destination):
-                    ip_address = destination[0]
-                else:
+                def on_error(failure):
+                    self.tunnel_logger.error("Can't resolve ip address for hostname %s. Failure: %s",
+                                             destination[0], failure)
+
+                def on_ip_address(ip_address):
                     try:
-                        ip_address = socket.gethostbyname(destination[0])
-                        self.tunnel_logger.info("Resolved ip address %s for hostname %s",
-                                                 ip_address,
-                                                 destination[0])
-                    except:
-                        self.tunnel_logger.error("Can't resolve ip address for hostname %s", destination[0])
-
-                try:
-                    self.transport.write(data, (ip_address, destination[1]))
-                except Exception, e:
-                    self.tunnel_logger.error("Failed to write data to transport: %s. Destination: %s",
+                        self.transport.write(data, (ip_address, destination[1]))
+                    except Exception, e:
+                        self.tunnel_logger.error("Failed to write data to transport: %s. Destination: %s",
                                              e[1],
                                              repr(destination))
-                    raise
+                        raise
+
+                    self.community.increase_bytes_sent(self, len(data))
 
-                self.community.increase_bytes_sent(self, len(data))
+                resolve_ip_address_deferred = self.resolve_ip_address_for_destination(destination)
+                resolve_ip_address_deferred.addCallbacks(on_ip_address, on_error)
+                # Add it to the deferred list to ensure it's cancelled when exiting.
+                self.deferred_list.append(resolve_ip_address_deferred)
+                # And add a callback to remove it if it fires so we don't keep it in memory.
+                resolve_ip_address_deferred.addBoth(lambda _: self.deferred_list.remove(resolve_ip_address_deferred))
```

Comment:
This should pass through the result, otherwise it swallows any error that may happen upstream.